### PR TITLE
Missing opt with ctlz and shifts of power of 2 constants (#41333)

### PR DESCRIPTION
**Pull request by snikitav**
This patch does the following simplifications:
```
cttz(shl(C, X), 1) -> add(cttz(C, 1), X)
cttz(lshr exact(C, X), 1) -> sub(cttz(C, 1), X)
ctlz(lshr(C, X), 1) --> add(ctlz(C, 1), X)
ctlz(shl nuw (C, X), 1) --> sub(ctlz(C, 1), X)
```
Alive2: https://alive2.llvm.org/ce/z/9KHlKc
Closes #41333

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Could you please add the alive2 proof to the PR description?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 1]
```diff
@@ -514,6 +514,8 @@ static Instruction *foldCttzCtlz(IntrinsicInst &II, InstCombinerImpl &IC) {
     return IC.replaceInstUsesWith(II, ConstantInt::getNullValue(II.getType()));
   }
 
+  ConstantInt *CInt;
```
[the remark this answers is not in the document]
Comment (snikitav) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:517]: If change `ConstantInt` to `Constant` not only constant splat vectors may be handled:
```llvm
----------------------------------------
define <2 x i32> @src(<2 x i32> %#0) {
#1:
  %div = lshr <2 x i32> { 8387584, 1234567 }, %#0
  %ctlz = ctlz <2 x i32> %div, 1
  ret <2 x i32> %ctlz
}
=>
define <2 x i32> @tgt(<2 x i32> %#0) {
#1:
  %ctlz = add <2 x i32> %#0, { 9, 11 }
  ret <2 x i32> %ctlz
}
Transformation seems to be correct!
```

File: llvm/test/Transforms/InstCombine/ctlz-cttz-shifts.ll [hunk 2]
```diff
@@ -0,0 +1,36 @@
+; RUN: opt -passes=instcombine -S < %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (snikitav) [reply llvm/test/Transforms/InstCombine/ctlz-cttz-shifts.ll:2]: done

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 3]
```diff
@@ -549,6 +551,36 @@ static Instruction *foldCttzCtlz(IntrinsicInst &II, InstCombinerImpl &IC) {
 
     if (match(Op0, m_Intrinsic<Intrinsic::abs>(m_Value(X))))
       return IC.replaceOperand(II, 0, X);
+
+    // cttz(shl(%const, %val), 1) --> add(cttz(%const, 1), %val)
+    if (match(Op0, m_Shl(m_ConstantInt(CInt), m_Value(X))) &&
+        match(Op1, m_One())) {
+      Value *ConstCttz =
+          IC.Builder.CreateBinaryIntrinsic(Intrinsic::cttz, CInt, Op1);
+      return IC.replaceInstUsesWith(II, IC.Builder.CreateAdd(ConstCttz, X));
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp]: Inferring `nsw/nuw` flags can exploit more optimization opportunities. You don't need to do anything until we find it is a missed optimization in some real-world applications.

But if you think it has value I'm OK with it. Beware of the type as `nsw/nuw` flags may not be applicable to `i1/i2` :)

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 4]
```diff
@@ -549,6 +551,37 @@ static Instruction *foldCttzCtlz(IntrinsicInst &II, InstCombinerImpl &IC) {
...
+      Value *ConstCtlz =
+          IC.Builder.CreateBinaryIntrinsic(Intrinsic::ctlz, C, Op1);
+      return BinaryOperator::CreateAdd(ConstCtlz, X);
+    }
+
+    // ctlz(shl nuw (%const, %val), 0)     |
+    // ctlz(shl nsw (%const, %val), 0)     |--> sub(ctlz(%const, 0), %val)
+    // ctlz(shl nuw nsw (%const, %val), 0) |
+    if (match(Op0, m_NUWShl(m_Constant(C), m_Value(X))) ||
+        match(Op0, m_NSWShl(m_Constant(C), m_Value(X)))) {
```
[the remark this answers is not in the document]
Comment (snikitav) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp]: Thank you for pointing this out. Fixed the code. BTW, is there a comprehensive guide how to use alive2 properly?

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Review by nikic — APPROVED** [review]
LGTM apart from one issue.

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 5]
```diff
@@ -549,6 +551,36 @@ static Instruction *foldCttzCtlz(IntrinsicInst &II, InstCombinerImpl &IC) {
 
     if (match(Op0, m_Intrinsic<Intrinsic::abs>(m_Value(X))))
       return IC.replaceOperand(II, 0, X);
+
+    // cttz(shl(%const, %val), 1) --> add(cttz(%const, 1), %val)
+    if (match(Op0, m_Shl(m_Constant(C), m_Value(X))) && match(Op1, m_One())) {
```
[the remark this answers is not in the document]
Comment (snikitav) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp]: Fixed

**Review by dtcxzyw — APPROVED** [review]

**Pull request merged**